# 1. Global variables

In [ ]:
# x = 5

# def f(alist):
#     for i in range(x):
#         alist.append(i)
#     return alist

# alist = [1, 2, 3]
# ans = f(alist)
# print(ans)
# print(alist) # alist has been changed


def f(alist):
  x = 5
  returnlist = alist[0:len(alist)]
  for i in range(x):
      returnlist.append(i)
  return returnlist

alist = [1, 2, 3]
ans = f(alist)
print(ans)
print(alist) # alist has not been changed


[1, 2, 3, 0, 1, 2, 3, 4]
[1, 2, 3]


# 2. List comprehension

In [1]:
ans = list(map(lambda x: x * x, filter(lambda x: x % 2 == 1, range(10))))

print(ans)

ans_comprehension = [x*x for x in filter(lambda x: x % 2 == 1, range(10))]

print(ans_comprehension)

[1, 9, 25, 49, 81]
[1, 9, 25, 49, 81]


# 3. Filter list

In [ ]:
ilist = input().split(' ')
n = int(input())

def smaller_than(wordd):

    return len(wordd) <= n

ans = list(filter(smaller_than, ilist))

print(ans)

test the words for lengths smaller than four
4
['test', 'the', 'for', 'than', 'four']


# 4.  Map dictionary

In [ ]:
lang = {"Python" : 3, "Java" : '', "Cplusplus" : 'test', "Php" : 0.7}

def returnKeyLengths(dic):
  return list(map(len, dic.keys()))

print(returnKeyLengths(lang))

[6, 4, 9, 3]


# 5. Lambda functions



In [ ]:
language_scores = [('Python', 97), ('Cplusplus', 81), ('Php', 45), ('Java', 32)]

language_scores.sort( key = lambda x : x[0] )

print(language_scores)

[('Cplusplus', 81), ('Java', 32), ('Php', 45), ('Python', 97)]


# 6. Nested functions



In [ ]:
def squared ( x ):

  return x * x

def cubed ( x ):

  return x * x * x

def sixthPower ( x ):

  return(squared(cubed(x)))


#eg:
i = int(input())

print(sixthPower(i))

2
64


# 7. Decorators



In [ ]:
def hello(func):
    def wrapper(i):
        print("Hello World!")
        return func(i)
    return wrapper

@hello
def square(x):
    return x*x

#example
i = int(input())
print(square(i))

2
Hello World!
4


# 8. The Fibonacci sequence (part 2)



In [ ]:
def fibb(x):
    if x == 1 or x == 2:
        return 1
    else:
        # recursive call to the function
        return fibb(x - 2) + fibb(x - 1)

for i in range(1, 21):
  print(fibb(i))

1
1
2
3
5
8
13
21
34
55
89
144
233
377
610
987
1597
2584
4181
6765


# 9. The Fibonacci sequence (part 3)



In [ ]:
def loopFib(x):
  fib = [1, 1]

  for i in range(x):
    fib.append(fib[i] + fib[i+1])

  return fib

def recursiveFib(x):
   fibonacci = []
   if x == 1 or x == 2:
     return 1
   else:
       # recursive call to the function
       return fibb(x - 2) + fibb(x - 1)


%timeit loopFib(20)

%timeit recursiveFib(20)

3.48 µs ± 144 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)
1.98 ms ± 40.5 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)


the first approach is obviusly the more efficient application because the time for calculating the 20th fibonacci is 3.48 microseconds compared to the 1.98 miliseconds it took the recursive function to calculate, which is approximately $1.75 * 10^3$ times faster

# 10. Class definition



In [32]:
class Polygon:
  sides = []

  def getlengths(self, vals):
    self.sides = list(vals.split(','))
 
  def perimeter(self):
    perimeterr = 0
    for i in self.sides:
      perimeterr += int(i)

    return perimeterr

  def getOrderedSides(self, increasing):
    if increasing:
      return sorted(self.sides, key = lambda x: x[0])
    else:
      return sorted(self.sides, key = lambda x: x[0], reverse = True)

example = Polygon()
example.getlengths("3,5,4")
print(example.perimeter())
print(example.getOrderedSides(True))
print(example.getOrderedSides(False))

12
['3', '4', '5']
['5', '4', '3']


#11. Class inheritance



In [38]:
class rectangle(Polygon):

  isRect = True

  def getlengths(self, vals):
    self.sides = list(vals.split(','))
    if len(self.sides) != 2:
      print("this is not a rectangle")
      self.isRect = False
    
  
  def area(self):
    if self.isRect:
      area = int(self.sides[0]) * int(self.sides[1])
      return area
    else:
      print("this is not a rectangle")

rect = rectangle()
rect.getlengths("2,23")
print(rect.area())

not_rect = rectangle()
not_rect.getlengths("3,4,5")
not_rect.area()

46
this is not a rectangle
this is not a rectangle
